### PR TITLE
NEW: Add libcudss recipe

### PR DESCRIPTION
**Pull request by carterbox**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [ ] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by carterbox** [comment]
@adibbley, could you please take a look at this recipe? I'm my first time writing a recipe for a pre-compiled CUDA library.

File: recipes/libcudss/meta.yaml
Comment (adibbley) [inline recipes/libcudss/meta.yaml:33]: For Windows we've been using the `LIBRARY_LIB` and `LIBRARY_BIN` locations rather than `PREFIX`
```suggestion
    - cp -rv lib %LIBRARY_LIB%  # [win]
    - cp -rv bin %LIBRARY_BIN%  # [win]
```

File: recipes/libcudss/meta.yaml
Comment (adibbley) [inline recipes/libcudss/meta.yaml:70]:
```suggestion
        - if not exist %LIBRARY_BIN%\\cudss64_0.dll exit 1  # [win]
```

File: recipes/libcudss/meta.yaml
Comment (adibbley) [inline recipes/libcudss/meta.yaml:115]:
```suggestion
        - if not exist %LIBRARY_LIB%\\cudss.lib exit 1  # [win]
        - if not exist %LIBRARY_INC%\\cudss.h exit 1  # [win]
        - if not exist %LIBRARY_INC%\\cudss_distributed_interface.h exit 1  # [win]
```

File: recipes/libcudss/meta.yaml
Comment (adibbley) [inline recipes/libcudss/meta.yaml:31]:
```suggestion
    - cp -rv include %LIBRARY_INC%  # [win]
```

File: recipes/libcudss/meta.yaml
Comment (adibbley) [inline recipes/libcudss/meta.yaml:110]:
```suggestion
        - test -f $PREFIX/include/cudss.h  # [linux]
        - test -f $PREFIX/include/cudss_distributed_interface.h  #[linux]
```

**Review by adibbley — COMMENTED** [review]
Thanks, Daniel! Overall looks good to me, just some comments on Windows locations.

cc @billysuh7 for viz

**Comment by carterbox** [comment]
Thanks, @adibbley for the for those suggestions!

File: recipes/libcudss/meta.yaml
Comment (carterbox) [inline recipes/libcudss/meta.yaml:58]: I also want some feedback from @kvoronin just to check that the dependencies are correct, and that they aren't any missing deps. cublas, for example.

File: recipes/libcudss/meta.yaml
Comment (kvoronin) [inline recipes/libcudss/meta.yaml:42]: Do we (usually) ship cmake config/version files via conda? Or only in libcudss-dev?
What about include/?

File: recipes/libcudss/meta.yaml
Comment (kvoronin) [inline recipes/libcudss/meta.yaml:53]: If we use `arm-variant`  and it all works, can we include the jetson build into the conda distribution? We currently have a special build for jetson/Orin for cudss (starting with cudss 0.3.0)

File: recipes/libcudss/meta.yaml
Comment (kvoronin) [inline recipes/libcudss/meta.yaml:58]: nccl and openmpi are only required if customers want to use MGMN mode of cudss. I don't like imposing this requirement on customers who will only use cudss on a single GPU (and never enable MGMN mode). Or WDYT, is having an extra dependency for them is fine? Like there is no nccl for Windows, will it be a problem for using conda package of cudss on Windows?

File: recipes/libcudss/meta.yaml
Comment (kvoronin) [inline recipes/libcudss/meta.yaml:82]: A reminder for myself to clean the wording here, I believe it can be improved.

**Comment by kvoronin** [comment]
Will we have a `run_test.<sh|bat>` like we do for other  CUDA math libraries?

File: recipes/libcudss/meta.yaml
Comment (kvoronin) [inline recipes/libcudss/meta.yaml:156]: I also wonder if it makes sense to get one of the cudss developers (me) into the maintainers list.

File: recipes/libcudss/meta.yaml
Comment (carterbox) [reply recipes/libcudss/meta.yaml:47]: The Windows platform requires the import library at compile time in order to create dynamic links to the shared runtime library (`*.dll`), so without it, no one will be able to link to this package.

Comment (carterbox) [reply recipes/libcudss/meta.yaml:47]: Import libraries are not required at runtime which is why it is being shipped in the `-dev` package.

File: recipes/libcudss/meta.yaml
Comment (carterbox) [reply recipes/libcudss/meta.yaml:58]: Good catch! We need to guard `nccl` with a platform selector. We can move the files which link to openmpi to a separate output, so users don't get them by default. Probably we will try to have a nompi build variant in order to match the rest of the channel.

File: recipes/libcudss/meta.yaml
Comment (carterbox) [reply recipes/libcudss/meta.yaml:156]:
```suggestion
    - conda-forge/cuda
    - kvoronin
```

File: recipes/libcudss/meta.yaml
Comment (carterbox) [reply recipes/libcudss/meta.yaml:53]: I think we can, but I'm not sure if Conda can select for the user the tegra variant automatically yet. @jakirkham would know more about that.

File: recipes/libcudss/meta.yaml
Comment (carterbox) [reply recipes/libcudss/meta.yaml:42]: This recipe creates two outputs: `libcudss`, and `libcudss-dev`. The non-dev package contains only artifacts needed at runtime. The dev package contains only things needed at compile time.

We typically ship everything via conda except for documentation, tests, and examples because end users are mostly not developers and people use the internet for documentation and examples.

**Comment by carterbox** [comment]
> Will we have a `run_test.<sh|bat>` like we do for other CUDA math libraries?

There are no GPUs on the build nodes available to run tests. Did you have something besides existence that you wanted to test? We can do a test build with CMAKE to check that the config is valid.

**Comment by kvoronin** [comment]
> > Will we have a `run_test.<sh|bat>` like we do for other CUDA math libraries?
> 
> There are no GPUs on the build nodes available to run tests. Did you have something besides existence that you wanted to test? We can do a test build with CMAKE to check that the config is valid.

I am thinking about a test like for cusparselt:https://github.com/conda-forge/cusparselt-feedstock/blob/main/recipe/run_test.sh or similar. I know there was a general question raised as to how we should write such simple tests across the CUDA math libraries. So maybe there is a different idea. @jakirkham should know more about this.

File: recipes/libcudss/meta.yaml
Comment (carterbox) [reply recipes/libcudss/meta.yaml:53]: There is no support for tegra variant yet, but this pattern is a template that should allow it in the future.

File: recipes/libcudss/meta.yaml
Comment (carterbox) [reply recipes/libcudss/meta.yaml:58]: @kvoronin, do downstream packages which link against the nccl or openmpi comm layers need to also link against the same nccl/openmpi directly (they are public linkages)? Or are these private linkages which are abstracted away by the cudss API?

Either way, I need to know what versions of openmpi and nccl you link against, so that the recipe can have minimum versions of nccl and openmpi.

File: recipes/libcudss/test/CMakeLists.txt
Comment (carterbox) [inline recipes/libcudss/test/CMakeLists.txt:7]: @kvoronin, I added a compiler test, but it fails because `libcudss_static.a` is missing (as per channel policy we don't ship static libraries). This needs to be fixed upstream. cudss should allow cmake to find the cudss package without finding the static library (the static library should be an optional component).

File: recipes/libcudss/meta.yaml
Comment (carterbox) [reply recipes/libcudss/meta.yaml]: I was looking at the headers because of making the compile test and it seems that the API does not mention nccl and openmpi, so whether to use either of these libraries is decided at runtime?

File: recipes/libcudss/meta.yaml
Comment (kvoronin) [inline recipes/libcudss/meta.yaml:186]: Suggestion is to replace this description with:
```
    NVIDIA cuDSS is an optimized, first-generation
    GPU-accelerated Direct Sparse Solver library for solving linear systems
    with sparse matrices. Direct Sparse Solvers are an important part
    of numerical computing as they provide a general robust way of solving large
    linear systems without and are capable of taking advantage of both high compute
    throughput and memory bandwidth of the GPUs.
```

File: recipes/libcudss/meta.yaml
Comment (kvoronin) [inline recipes/libcudss/meta.yaml:184]:
```suggestion
    with sparse matrices.
```

File: recipes/libcudss/meta.yaml
Comment (kvoronin) [inline recipes/libcudss/meta.yaml:140]: I've learned recently about conda having an mpi lock ( = single mpi per environment). Will `libcudss-commlayer-mpi` be takling advantage of that? E.g., uf the user first installs an incompatible MPI, will we be able to report smth if the user attempt later to install `libcudss-commlayer-mpi`?

File: recipes/libcudss/meta.yaml
Comment (carterbox) [inline recipes/libcudss/meta.yaml:191]:
```suggestion
  description: >-
    NVIDIA cuDSS is an optimized, first-generation
    GPU-accelerated Direct Sparse Solver library for solving linear systems
    with sparse matrices. Direct Sparse Solvers are an important part
    of numerical computing as they provide a general robust way of solving large
    linear systems without and are capable of taking advantage of both high compute
    throughput and memory bandwidth of the GPUs.
```

File: recipes/libcudss/meta.yaml
Comment (carterbox) [inline recipes/libcudss/meta.yaml:129]:
```suggestion
        - nccl  # collect channel pinnings
        - nccl >=2.0,<3
```

File: recipes/libcudss/meta.yaml
Comment (carterbox) [reply recipes/libcudss/meta.yaml:128]: I need to know which minor version CUDSS is built against too. Otherwise, it is possible that CUDSS is compiled against a new NCCL before NCCL is released on conda-forge and then the NCCL binary will be missing a feature that CUDSS needs.

File: recipes/libcudss/meta.yaml
Comment (kvoronin) [reply recipes/libcudss/meta.yaml:74]: I treat the src/ more like "auxiiliary" than as "examples" (because we do have pure examples which show how to use the library).
Do I get it right that you suggest to do one more package so that to get these src/ the user would do `conda install libcudss, libcudss-examples` or di you mean smth else?
We can expect a significant amount of users who would want to build their own communication layer so ideally I'd have it available "by default" if it is allowed by the common practices.

File: recipes/libcudss/meta.yaml
Comment (kvoronin) [reply recipes/libcudss/meta.yaml:140]: It was NCCL 2.21(.5). I think we don't use any features above the major version 2.0.

Comment (kvoronin) [reply recipes/libcudss/meta.yaml:140]: But i don't think we rely on any super-new features of NCCL (hopefully using an older version will not run into the fixed bugs since the last official release).

File: recipes/libcudss/meta.yaml
Comment (carterbox) [reply recipes/libcudss/meta.yaml:155]: The `libcudss-commlayer-mpi` package requires openmpi, so conda will switch to openmpi if another mpi is installed in the environment or it will raise an error because it is not able to find an mpi that is compatible with all the requested packages.

**Comment by carterbox** [comment]
I believe the only thing remaining is the upstream patch for building without static libraries installed.

**Comment by carterbox** [comment]
Waiting for next release of cudss in December 2024.

**Comment by carterbox** [comment]
@conda-forge/help-c-cpp, please review

**Review by kvoronin — APPROVED** [review]
LGTM

**Review by adibbley — APPROVED** [review]

**Pull request merged**
